server/src/main/java/org/elasticsearch/search/aggregations/metrics/WrapperTDigestArrays.java

```diff
@@ -0,0 +1,365 @@
+
+    private final CircuitBreaker breaker;
+
+    public WrapperTDigestArrays(CircuitBreaker breaker) {
+        this.breaker = breaker;
+    }
+
+    @Override
+    public WrapperTDigestDoubleArray newDoubleArray(int initialCapacity) {
+        return validate(new WrapperTDigestDoubleArray(breaker, initialCapacity));
+    }
+
+    @Override
+    public WrapperTDigestIntArray newIntArray(int initialSize) {
+        return validate(new WrapperTDigestIntArray(breaker, initialSize));
+    }
+
+    @Override
+    public TDigestLongArray newLongArray(int initialSize) {
+        return validate(new WrapperTDigestLongArray(breaker, initialSize));
+    }
+
+    @Override
+    public TDigestByteArray newByteArray(int initialSize) {
+        return validate(new WrapperTDigestByteArray(breaker, initialSize));
+    }
+
+    public WrapperTDigestDoubleArray newDoubleArray(double[] array) {
+        return validate(new WrapperTDigestDoubleArray(breaker, array));
+    }
+
+    public WrapperTDigestIntArray newIntArray(int[] array) {
+        return validate(new WrapperTDigestIntArray(breaker, array));
+    }
+
+    /**
+     * Safely adjusts the breaker with an array.
+     * <p>
+     *     Copied from {@link org.elasticsearch.common.util.BigArrays}.
+     * </p>
+     */
+    private <T extends Accountable & Releasable> T validate(T array) {
+        boolean success = false;
+        try {
+            adjustBreaker(array.ramBytesUsed());
+            success = true;
+        } finally {
+            if (success == false) {
+                Releasables.closeExpectNoException(array);
+            }
+        }
+        return array;
+    }
+
+    /**
+     * Adjusts the breaker with the delta for an already created array.
+     * <p>
+     *     Copied from {@link org.elasticsearch.common.util.BigArrays}.
+     * </p>
+     */
+    void adjustBreaker(final long delta) {
+        // checking breaker means potentially tripping, but it doesn't
+        // have to if the delta is negative
+        if (delta > 0) {
+            try {
+                breaker.addEstimateBytesAndMaybeBreak(delta, "tdigest-array-wrapper");
+            } catch (CircuitBreakingException e) {
+                // since we've already created the data, we need to
+                // add it so closing the stream re-adjusts properly
+                breaker.addWithoutBreaking(delta);
+                throw e;
+            }
+        } else {
+            breaker.addWithoutBreaking(delta);
+        }
+    }
+
+    private abstract static class MemoryAwareArray implements Releasable, Accountable {
+        protected final CircuitBreaker breaker;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        MemoryAwareArray(CircuitBreaker breaker) {
+            this.breaker = breaker;
+        }
+
+        @Override
+        public final void close() {
+            if (closed.compareAndSet(false, true)) {
```

Comment:
What are the scenarios where this is called repeatedly? Maybe document them in a comment?

Reply:
No "expected" scenario, but the `Closeable` interface expects no effect on multiple calls to `close()`.
Personally, I wouldn't comment it, as many other classes use this mechanism, and otherwise we would be commenting every "close" with Java expected behaviours.
That said, if you really think it makes sense, I have no problem in adding it

Reply:
Right, I noticed too that this is a standard pattern. Sounds good.